### DISCUSSION TITLE
Remote Audio Transcription and Idle Client

### DISCUSSION THREAD

**Discussion by gfd2020** [sepinf-inc/IPED#2001] [Ideas]
Remote audio transcription with the IPED server/client is excellent. It helps a lot in this difficult task.
However, I noticed something that could perhaps improve performance a little. 
For example: a UFDR report has photos, spreadsheets, chats, audio, etc. When the client identifies audio it will send it to the remote server. But sometimes it uses all threads to send to the server, so it remains idle and waiting for the result. Perhaps we could configure a maximum number of transcription threads if it have other types of files not yet processed (maybe half), so as not to leave the client inactive and doing other tasks.

**Comment by lfcnassif** [comment]
Hi @gfd2020, that's a good idea! Actually we do a similar thing with simultaneous containers expansion, at some point we limited it to half the number of CPU cores, other containers are re-enqueued. We could do something similar to audios.

But this should be Configurable. If your transcription cluster is slow, clients could be very idle and could do other things. If the cluster is very fast, our case, decreasing the number of simultaneous transcription requests can waste the cluster resources and may decrease the overall throughput. To know if there are other types of files different than audios in the queue is another challenge, we don't have this info today, not sure how to do this...

**Reply by gfd2020** [reply]
I see. You are right about the challengs.
I think we can do a heuristic to see the number of audiotranscript threads dynamically change. We would base it on the seconds of the audio file, the average transcription time that the server processes, and the number of items remaining, etc.

I did a little test here ( My computer has 12 workers). It's not ideal, but just to see if it works. I create a static variable 'num' in the AbstractTranscriptTask class. Inside the process method I use it to know how many tasks are executing, and I control it to not exceed 4 ( if does I just call super.reEnqueueItem(evidence) ). In the RemoteWav2Vec2TranscriptTask class, I decrement this variable when a transcription is finished. This way, only 4 audio transcriptions are active. This way, I managed to ensure that my client never stands still.
Otherwise in the normal implementation, when a lot of audio arrives, it stays still.

**Reply by lfcnassif** [reply]
And overall processing time decreased in your case, right?

Maybe we can use a simpler heuristic: use half the workers for remote transcription initially and re-enqueue excedent audios, marking them as re-enqueued and incrementing a reEnqueued audios counter (if not re-enqueued before). If the number of items in the queue is equal or less then the number of re-enqueued audios, start using all threads for remote transcription. And just before transcribing an audio, check its re-enqueued flag and decrement the counter accordingly.

**Reply by gfd2020** [reply]
> And overall processing time decreased in your case, right?

Yes, exactly.

> Maybe we can use a simpler heuristic: use half the workers for remote transcription initially and re-enqueue excedent audios, marking them as re-enqueued and incrementing a reEnqueued audios counter (if not re-enqueued before). If the number of items in the queue is equal or less then the number of re-enqueued audios, start using all threads for remote transcription. And just before transcribing an audio, check its re-enqueued flag and decrement the counter accordingly.

Okay, I'll test your suggestion.

I also thought about splitting the heuristic into two parts. The first is to define a minimum number of workers for the remote task, based on its speed, and changed it dynamically.
The second part is where to put the item in the queue. The default ReQueue sends it to the end of the queue, but I don't think it's the best option.
I made a sprayed requeue.
If an audio has to be requeued, it will go to position 1/4 of the queue, if another audio enters the requeue in less than 5 seconds, it is positioned in the position of the previous one + one space in the number of workers. ( ex.: lastQueuePlace + 12 ) 
If it is longer than 5 seconds, then this audio will enter 1/4 of the queue again.

With this algorithm, the tests I did were very good ( not perfect ), my client didn't spend 1 second without work and the server always had at least one task. In this test, I fixed a maximum of 8 tasks for transcription on a total of 12 workers per processor.

**Reply by lfcnassif** [reply]
Those are very good ideias. I thought about a dynamic audio retry interval when the service is busy when implementing it initially, but decided to keep it simple. Anyway it is a different aspect of the feature.

**Reply by gfd2020** [reply]
I'm still trying other things. I have a question. I think you have a server with some RTX 3090, right? I imagine it must be quick. How long does a transcription of a 30 seconds audio last ( in wav2vec2 ) ? 1 second on average?

**Reply by lfcnassif** [reply]
Yes, on RTX 3090 the large wav2vec2 model (1B parameters) does the job about 40-50x faster then the audio duration and we are able to run 2 simultaneous jobs per GPU. Today we have 8 servers with 2 GPUs each. That gives up to 1500x faster than audio real time when transcribing many small audios, today one large audio is not broken in separate jobs, which is fine for our current use case.

**Reply by lfcnassif** [reply]
So our clients are iddle for short periods of time. Of course, we can improve the feature for those without such a great GPU cluster infrastructure.

**Reply by gfd2020** [reply]
Hi @lfcnassif .
I've finished the heuristic (which can be turned on or off and configurable) and I'm going to open a new PR so that other developers can test and refine it, ok?

**Reply by lfcnassif** [reply]
Great! Thank you @gfd2020 for this optimization, it should help many users!
